### PR TITLE
[WIP] Feature/profile page - requests for comments

### PR DESCRIPTION
**Pull request by ccoenen**
I am currently implementing a profile page that would let you edit your display name (henceforth: `name`) and possibly edit your password. It is far from finished, and I did a little bit of groundwork for future pages along the way. Those should get their own pull requests, probably. Also this will contain commits _not_ directly related to either one, more specifically those from PR #947 )

Questions:
* with more and more pages, the "old" way of putting all the variables in all the render calls does not work any more. Express has `app.locals` which are available in the template files. Does this work for you? I find this much cleaner. All those changes are now proposed as PR in #1035
* Did I get all the places, where those variables are used? Does EJS have a way to detect used but unset variables (like a strict mode for EJS, perhaps)?
* We could put "delete user" on that page instead of the index-page-menu. Opinions?

TODO:
* [x] The header contains everything and the kitchen sink from the note editor. This needs to be fixed
* [x] user name will have to be sanitized. I would hate to restrict it to [a-z] (even though this would strictly not be a regression, it does feel like one). I have no experience with express/sequelize/ejs, so I would value your input - even just in form of documentation or a how-to, so I don't introduce XSS or something like that.
* [x] the password change is not implemented at all.
* [ ] make sure I didn't mess up any of the login providers along the way (i.e. disable password boxes for ldap or github ...)

`-----8<---- MVP line, cut here if neccessary ---8<----`

* [x] (rudimentary) visual styling (right now, this is intended for code critique and conceptual work. Styling will be done later, of course.)
* [x] translations should be added to the correct jsons
* [ ] username + displayname (display name defaulting to username, username defaulting to whatever it would have been so far. Perhaps the profile json will no longer be neccessary)
* [ ] have `/user/:username` actually use username instead of user.id
* [x] ~~documents i contributed and document I own is a bit tricky, those should be opt-in since they make documents discoverable that would otherwise have remained hidden.~~ (checked off for "won't be done for now")

For a MVP we could "just not link to it, yet" but already deploy it. We could get some feedback this way and improve it more before we officially ship it.

### CODE REVIEW

**Comment by SISheogorath** [comment]
> We could put "delete user" on that page instead of the index-page-menu. Opinions?

Definitely! It was only put into the menu on the main page, because there was no profile page yet.

> Did I get all the places, where those variables are used? Does EJS have a way to detect used but unset variables (like a strict mode for EJS, perhaps)?

Not sure if you got all, but EJS will definitely complain when there is something missing.

---

All in all, I really like your work! Seems like user profiles are moving forward! And I'm really curious about what we will get as result!

**Comment by davidmehren** [comment]
How does this interact with users from other auth sources, eg. LDAP? Normal apps might not have permission to change user attributes or delete them.

**Comment by SISheogorath** [comment]
@davidmehren right now, we delete all user data that exist in CodiMD's database when someone requests an account deletion. This means they can definitely re-auth with their external credentials afterwards, but they start from scratch, like their account never existed before.

**Comment by ccoenen** [comment]
Thanks for reminding me, until now i was only concerned with local users. I will add it to the TODO list. Since I use none of the other login providers, I'd value any input on that part. I might need to disable password change for those users, but display name (`name` attribute) ist completely local in any case. It's what will be listed in the "2 ONLINE" button top right, for example.

File: lib/models/user.js [hunk 1]
```diff
@@ -19,7 +19,11 @@ module.exports = function (sequelize, DataTypes) {
       unique: true
     },
     name: {
-      type: DataTypes.STRING
+      type: DataTypes.STRING,
+      validate: {
+        is: /^[a-z0-9\-+ ]+$/i,
```
[the remark this answers is not in the document]
Comment (ccoenen) [reply lib/models/user.js]: the name attribute is meant as display name. I'll change the property (and db field) name to reflect that. But even with that in place, I'd like to disallow at least `<>'"=%/:`

currently, we don't have a username. For local users, the email address would be the closest thing. For remotely authenticated, I have no clue what would be in its stead.

I wouldn't want to create a third thing (email - username - displayname) as we have not technical neccessity for it. We _could_ add something like a shorthand. Like John Doe -> johnd that could (one day) be used for things like `@johnd` (which I would like to link back to the user profile) or in urls (think `/profile/johnd` or perhaps `/johnd/note-1`).

[the remark this answers is not in the document]
Comment (ccoenen) [reply lib/models/user.js]: we could of course introduce (technical) username and (human readable) display name and

- make email optional (for now we're not using it anyway)
- make display name default to username, so people don't have to provide five thins when signing up.

[the remark this answers is not in the document]
Comment (SISheogorath) [reply lib/models/user.js]: Yes, that should be the way to go. We already have the concept of display names in various (maybe all except local?) login methods, so I think we should do this.

And while the username can be quite strict, the Display name allows all kinds of versions. And of course, yes, the default display name should simply be the username.

File: lib/models/user.js [hunk 2]
```diff
@@ -22,9 +22,10 @@ module.exports = function (sequelize, DataTypes) {
       type: DataTypes.STRING,
       validate: {
         is: /^[a-z0-9\-+ ]+$/i,
-        len: [1,255]
+        len: [2, 255]
```
Comment (SISheogorath) [inline lib/models/user.js]: I would even go for 3 characters as minimum length

File: locales/en.json [hunk 3]
```diff
@@ -1,117 +1,135 @@
...
+  "This will delete your account, all notes that are owned by you and remove all references to your account from other notes.": "This will delete your account, all notes that are owned by you and remove all references to your account from other notes.",
+  "Delete user": "Delete user",
+  "Export user data": "Export user data",
+  "Help us translating on %s": "Help us translating on %s",
+  "Account Settings for %s": "Account Settings for %s",
+  "Account Settings": "Account Settings",
+  "About You": "About You",
+  "Visit public profile": "Visit public profile",
+  "User could not be saved.": "User could not be saved",
+  "name/Validation is failed": "Name contains invalid characters",
```
Comment (ccoenen) [inline locales/en.json]: In two places I did not conform to the normal "human readable" pattern of this file. This is the first: the validation error messages from sequelize are just bad. To know which field is affected, I concatenate them with a slash.

File: locales/en.json [hunk 4]
```diff
@@ -1,117 +1,135 @@
...
+  "name/Validation len failed": "Name is too short",
+  "email/Validation isEmail failed": "Email is not a valid address",
+  "invalid_password_given/invalid old password": "Old password is incorrect",
+  "confirmation_matches/password and confirmation do not match": "Password and confirmation do not match",
+  "user.name": "name",
```
Comment (ccoenen) [inline locales/en.json]: this is the other place: I want to translate the user properties, and I think it is important that we know that this is about the user model.

**Comment by ccoenen** [comment]
I will be going for username / displayname next. Along the way I'll probably rebase this a few times, so previous discussion might end up folded up in github.

**Comment by ccoenen** [comment]
closing in favor of https://github.com/codimd/server/pull/34

**Comment by Choque** [comment]
Just wondering, has the profile page ever been implemented? I just spun up a CodiMD instance in docker and I can't see to find my profile page nor change the (auto-generated) username.

**Comment by ccoenen** [comment]
I no longer contribute to CodiMD, but I proposed some of this in https://github.com/hedgedoc/hedgedoc/pull/34 as mentioned above.

**Pull request closed without merging**